### ISSUE TITLE
Typed settings

### ISSUE THREAD

**Issue by nerdondon** [elastic/elasticsearch-js#926]
## 🚀 Feature Proposal

Add strong types around the `body` property of API calls.

## Motivation

Currently the typings take a generic parameter and assign that type to the body
```
export interface IndicesCreate<T = any> extends Generic {
  index: string;
  // ...snip... 
  body?: T;
}
```

The body here can actually take settings for index creation and mappings. These things are documented here: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html and https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html.

Currently, I would have to recreate the interface specified in the docs and pass it as a generic parameter:

```
interface EveryoneWouldNeedToDoThis {
  settings: {};
  mappings: {};
}

const createIndexResult = client.indices.create<EveryoneWouldNeedToDoThis>({
  index: indexName,
  body: {
    number_of_replicas
  }
});
```

The issue also lies in that everyone would also need to re-create these types in their own code in order to take advantage of type-safety. The API clearly expects a well-formed request. Would it be possible to add these common interfaces to the types?

**Comment by delvedor** [comment]
Hello!
I agree that having strong body types would be great, but we implemented the current solution because we don't have yet a spec that we can use for generating the request and response body types.
We also can't write all of them by hand, because it would be too hard to maintain them considering the number of APIs of Elasticsearch and the different major versions that we are maintaining.

The current solution is not ideal, but it mitigates the issue, while we are working on improving this side :)

**Comment by delvedor** [comment]
The new types has been shipped with 7.13, see the [documentation](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/typescript.html) to learn how to use them!
